docs/security.rst

```diff
@@ -0,0 +1,232 @@
+
+Encryption::
+
+    id = AUTHENTICATOR(id_key, data)
+    compressed = compress(data)
+
+    iv = reserve_iv()
+    encrypted = AES_256_CTR(enc_key, 8-null-bytes || iv, compressed)
+    authenticated = type-byte || AUTHENTICATOR(enc_hmac_key, encrypted) || iv || encrypted
+
+
+Decryption::
+
+    # Given: input *authenticated* data, possibly a *chunk-id* to assert
+    type-byte, mac, iv, encrypted = SPLIT(authenticated)
+
+    ASSERT(type-byte is correct)
+    ASSERT( CONSTANT_TIME_COMPARISON( mac, AUTHENTICATOR(enc_hmac_key, encrypted) ) )
+
+    decrypted = AES_256_CTR(enc_key, 8-null-bytes || iv, encrypted)
+    decompressed = decompress(decrypted)
+
+    ASSERT( CONSTANT_TIME_COMPARISON( chunk-id, AUTHENTICATOR(id_key, decompressed) ) )
+
+Offline key security
+--------------------
+
+Borg cannot secure the key material while it is running, because the keys
+are needed in plain to decrypt/encrypt repository objects.
+
+For offline storage of the encryption keys they are encrypted with a
+user-chosen passphrase. A 256 bit key encryption key (KEK) is derived
+from the passphrase using PBKDF2-HMAC-SHA256 with a random 256 bit
+salt which is then used to Encrypt-then-MAC a packed
+representation of the keys with AES-256-CTR with a constant
+initialization vector of 0. The resulting MAC is stored alongside the
+ciphertext and converted to base64.
+
+This base64 blob (commonly referred to as *keyblob*)
+is then stored in the key file or in the repository (keyfile and
+repokey modes respectively).
+
+This scheme, and specifically the use of a constant IV with the CTR
+mode, is secure because an identical passphrase will result in a
+different derived KEK for every encryption due to the salt.
+
+.. _Horton principle: https://en.wikipedia.org/wiki/Horton_Principle
+.. _HKDF: https://tools.ietf.org/html/rfc5869
+.. _length extension: https://en.wikipedia.org/wiki/Length_extension_attack
+
+Remote RPC protocol security
+============================
+
+.. note:: This section could use expansion.
+
+The RPC protocol is fundamentally based on msgpack'd messages exchanged
+over an encrypted SSH channel (the system's SSH client is used for this
+by piping data from/to it).
+
+This means that the authorization and transport security properties
+are inherited from SSH and the configuration of the SSH client
+and the SSH server. Therefore the remainder of this section
+will focus on the security of the RPC protocol within Borg.
+
+The assumed worst-case a server can inflict to a client is a
+denial of repository service.
+
+The situation were a server can
+create a general DoS on the client should be avoided, but
+might be possible by eg. forcing the client to allocate large
+amounts of memory to decode large messages. However, since we
+are not aware of any `"zip bomb"`_ equivalent in msgpack this
+would mean that the server has to send a similar amount of data to
+the client, which is not always practical. This is tracked with :issue:`2139`.
```

Comment:
Hmm, you're right, even in use_list=False mode it seems to allocate... something... before building the tuple:

```
>>> import msgpack
>>> u = msgpack.Unpacker(use_list=False)
>>> u.feed(b'\xdd\x7f\xff\xff\xff')
>>> next(iter(u))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "msgpack/_unpacker.pyx", line 469, in msgpack._unpacker.Unpacker.__next__ (msgpack/_unpacker.cpp:5459)
  File "msgpack/_unpacker.pyx", line 400, in msgpack._unpacker.Unpacker._unpack (msgpack/_unpacker.cpp:4483)
MemoryError
```

I dunno how that even works, perhaps that *is* the actual tuple and it somehow (dirtily?) changes the values of the tuple after-the-fact or something like that.